kernel/src/hil/eic.rs

```diff
@@ -0,0 +1,58 @@
+//! Interface for external interrupt controller.
+//! Interrupt can be configured as asynchronous to operate
+//! during deep sleep mode where the EIC clock is disabled.
+//!
+//! A basic use case:
+//! A user button is configured for falling edge trigger and async mode
+//!
+
+// Author: Josh Zhang <jiashuoz@cs.princeton.edu>
+// Last modified June 26, 2019
+
+/// Enum for selecting which edge to trigger interrupts on.
+#[derive(Debug)]
+pub enum InterruptMode {
+    RisingEdge,
+    FallingEdge,
+    HighLevel,
+    LowLevel,
+}
+
+/// Enum for enabling/disabling filter
+pub enum FilterMode {
```

Comment:
As a more general question: I've not encountered a software configurable interrupt filter before, but that could easily be my ignorance of the world. Do we think this is a SAM4L-specific thing (i.e. shouldn't be included as part of the HIL), or have others seen this elsewhere?

Reply:
I can't think of having seen this before, and the SAM4L definitely took a kitchen sink approach. However, in general this _seems_ like a feature that could exist on any chip. I think it makes sense to add a HIL interface, and see if it ever gets used in the future. My two cents.

Reply:
Oh wow I totally professored this. I now realize what you are referring to. Yeah...this seems like it would be very difficult or impossible to implement [fully] on other chips.